### ISSUE TITLE
False positive `reportIndexIssue` for typings in `v1.1.376` when `pyproject.toml` contains `[tool.pyright]` section (even if this section is empty)

### ISSUE THREAD

**Issue by chegoryu** [microsoft/pyright#8797]
**Describe the bug**

I gets false positive `Subscript for class "list" will generate runtime exception; enclose type annotation in quotes (reportIndexIssue)` error with pyright `v1.1.376`+ when  `[tool.pyright]` is present in `pyproject.toml` on this code (even if this section is empty):

```python
def f(_: int) -> list[int]:
    return []

```

Works fine with `v.1.1.375`:
```shell
$ cat pyproject.toml 
[tool.pyright]
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.375 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --project pyproject.toml 
WARNING: there is a new pyright version available (vv1.1.375 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 599ms

found 0 vulnerabilities
0 errors, 0 warnings, 0 informations
```

False positive error with `v1.1.376`:

```shell
$ cat pyproject.toml 
[tool.pyright]
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.376 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --project pyproject.toml 
WARNING: there is a new pyright version available (vv1.1.376 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 778ms

found 0 vulnerabilities
/home/echunaev/junk/test/test.py
  /home/echunaev/junk/test/test.py:1:18 - error: Subscript for class "list" will generate runtime exception; enclose type annotation in quotes (reportIndexIssue)
1 error, 0 warnings, 0 informations
```

Removing `[tool.pyright]` section solves the issue:

```shell
$ rm pyproject.toml && touch pyproject.toml
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.376 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --project pyproject.toml 
WARNING: there is a new pyright version available (vv1.1.376 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 604ms

found 0 vulnerabilities
0 errors, 0 warnings, 0 informations
```

Python version:

```shell
$ python3.11 --version
Python 3.11.7
```

**Comment by erictraut** [comment]
The error you're reporting is correct if the pythonVersion is 3.8 or older. That means pyright somehow thinks that it's being asked to assume Python 3.8. Let's see if we can figure out what that's the case.

If you manually add a `pythonVersion: "3.11"` to the config file, I'm guessing that will fix the issue. Could you confirm that theory?

**Comment by chegoryu** [comment]
Adding an explicit version fixes the issue:
```shell
$ cat pyproject.toml
[tools.pyright]
pythonVersion = "3.11"
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.376 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --project pyproject.toml
WARNING: there is a new pyright version available (vv1.1.376 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 516ms

found 0 vulnerabilities
0 errors, 0 warnings, 0 informations

```

**Comment by erictraut** [comment]
I'm guessing that you haven't configured your shell environment to use Python 3.11 by default. Pyright determines the default python version by running `python3 --version`. The recommended approach used by most pyright users is to activate a virtual environment in their shell before invoking pyright. This virtual environment will then become the default (target) Python environment, and pyright will get its import paths, pythonVersion, and pythonPlatform from this.

Which Python environment do you have configured in your shell?

@rchiodo, I'm guessing that this behavior is fallout from the recent change you made to the config code paths. It seems strange that the behavior changes here depending on whether there is a "[tools.pyright]" section present in the config file. Note that there are no language server settings involved.

**Comment by rchiodo** [comment]
There could be a bug with the default python version when a config file is present. Let me do some digging.

**Comment by erictraut** [comment]
> why do you use python3 --version when explicit --pythonpath /usr/bin/python3.11 is provided?

If you specify the pythonpath directly, we should use that binary. I didn't notice that you were using that option. We should fall back on "python3" only if there is no pythonpath configured.

**Comment by rchiodo** [comment]
I'm having trouble reproducing. The python version is always computed from the python path AFAICT. Config file has no impact on that.

**Comment by rchiodo** [comment]
@chegoryu can you run the error condition again with the '--verbose' flag and report the output?

**Comment by erictraut** [comment]
Unfortunately, the `--verbose` flag probably won't tell us much because of [this bug](https://github.com/microsoft/pyright/issues/8798) that I just fixed in the main branch.

**Comment by chegoryu** [comment]
```shell
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.375 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --verbose 
WARNING: there is a new pyright version available (vv1.1.375 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 620ms

found 0 vulnerabilities
Setting pythonPath for service "<default>": "/usr/bin/python3.11"
Loading pyproject.toml file at /home/echunaev/junk/test/pyproject.toml
Assuming Python version 3.11.7.final.0
No include entries specified; assuming /home/echunaev/junk/test
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Search paths for file:///home/echunaev/junk/test
  /home/echunaev/.cache/pyright-python/v1.1.375/node_modules/pyright/dist/typeshed-fallback/stdlib
  /home/echunaev/junk/test
  /home/echunaev/junk/test/typings
  /home/echunaev/.cache/pyright-python/v1.1.375/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/twix/python3.11/lib/python3.11
  /usr/twix/python3.11/lib/python3.11/lib-dynload
  /home/echunaev/.local/lib/python3.11/site-packages
  /usr/twix/python3.11/lib/python3.11/site-packages
Found 1 source file
pyright 1.1.375
0 errors, 0 warnings, 0 informations
Completed in 0.553sec



$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.376 python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --verbose   
WARNING: there is a new pyright version available (vv1.1.376 -> v1.1.377).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`


up to date, audited 2 packages in 609ms

found 0 vulnerabilities
Loading pyproject.toml file at /home/echunaev/junk/test/pyproject.toml
Assuming Python version 3.8.10.final.0
Setting pythonPath for service "<default>": "/usr/bin/python3.11"
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Search paths for file:///home/echunaev/junk/test
  /home/echunaev/.cache/pyright-python/v1.1.376/node_modules/pyright/dist/typeshed-fallback/stdlib
  /home/echunaev/junk/test
  /home/echunaev/junk/test/typings
  /home/echunaev/.cache/pyright-python/v1.1.376/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/twix/python3.11/lib/python3.11
  /usr/twix/python3.11/lib/python3.11/lib-dynload
  /home/echunaev/.local/lib/python3.11/site-packages
  /usr/twix/python3.11/lib/python3.11/site-packages
Found 1 source file
pyright 1.1.376
/home/echunaev/junk/test/test.py
  /home/echunaev/junk/test/test.py:1:18 - error: Subscript for class "list" will generate runtime exception; enclose type annotation in quotes (reportIndexIssue)
1 error, 0 warnings, 0 informations
Completed in 0.527sec
```

**Comment by chegoryu** [comment]
I see something like this with strace for `v1.1.376`: https://pastebin.com/KDskRkie
And this for `v1.1.375`: https://pastebin.com/99KmaCiV

Command is
```shell
$ PYRIGHT_PYTHON_FORCE_VERSION=v1.1.375 strace -F python3.11 -m pyright --pythonpath /usr/bin/python3.11 test.py --verbose 2>&1 | grep -e 'exec\|Assuming'
```

**Comment by rchiodo** [comment]
This output makes me think we don't recognize the pythonPath when the pyproject.toml is present for some reason:

```
[pid 82452] execve("/home/echunaev/.cargo/bin/python3", ["python3", "-I", "-c", "import os, os.path, sys; normali"...], 0x7f0cc1473d90 /* 53 vars */) = -1 ENOENT (No such file or directory)
[pid 82452] execve("/usr/local/sbin/python3", ["python3", "-I", "-c", "import os, os.path, sys; normali"...], 0x7f0cc1473d90 /* 53 vars */) = -1 ENOENT (No such file or directory)
[pid 82452] execve("/usr/local/bin/python3", ["python3", "-I", "-c", "import os, os.path, sys; normali"...], 0x7f0cc1473d90 /* 53 vars */) = -1 ENOENT (No such file or directory)
[pid 82452] execve("/usr/sbin/python3", ["python3", "-I", "-c", "import os, os.path, sys; normali"...], 0x7f0cc1473d90 /* 53 vars */) = -1 ENOENT (No such file or directory)
[pid 82452] execve("/usr/bin/python3", ["python3", "-I", "-c", "import os, os.path, sys; normali"...], 0x7f0cc1473d90 /* 53 vars */ <unfinished ...>
```

That looks like it's searching for a python3 match.

**Comment by rchiodo** [comment]
And these two messages are out of order:

```
Assuming Python version 3.8.10.final.0
Setting pythonPath for service "<default>": "/usr/bin/python3.11"
```

Setting the pythonPath always comes first no matter how I run it.

**Comment by chegoryu** [comment]
Shouldn't it be `configSettings` here?

https://github.com/microsoft/pyright/compare/1.1.375...1.1.376#diff-0a6cab623051b661e7b7f71bb6db134b1f37123b36e447c69fffd51baf5b937eR130

Because here it is `configSettings`:

https://github.com/microsoft/pyright/compare/1.1.375...1.1.376#diff-c3f0fa903325059f6e7dda1116458aafa72061ac3d1a51ddfd077a5dc0801030R326

**Comment by rchiodo** [comment]
@chegoryu this should be fixed in the next version.

**Comment by chegoryu** [comment]
Thank you so much.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.378.
